fiddler/pyproject.toml

```diff
@@ -0,0 +1,63 @@
+[build-system]
+requires = [
+    "hatchling>=0.13.0",
+    "setuptools<61",
+    "fiddler-client>=1.0.6",
+]
+build-backend = "hatchling.build"
+
+[project]
+name = "datadog-fiddler"
+description = "The fiddler check"
+readme = "README.md"
+license = {text = "BSD-3-Clause"}
+keywords = [
+    "datadog",
+    "datadog agent",
+    "datadog check",
+    "fiddler",
+]
+authors = [
+    { name = "Rajeev", email = "rajeev@fiddler.ai" },
+]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: System Administrators",
+    "License :: OSI Approved :: BSD License",
+    "Programming Language :: Python :: 2.7",
+    "Programming Language :: Python :: 3.8",
+    "Topic :: System :: Monitoring",
+]
+dependencies = [
+    "datadog-checks-base>=25.1.0",
+]
+dynamic = [
+    "version",
+]
+
+[project.optional-dependencies]
+deps = []
```

Comment:
Add fiddler-client==X.Y.Z

Reply:
@jose-manuel-almaza - I have made the changes to include the exact version of fiddler-client in the pyproject.toml file in three places. It still is giving me the same error. The three places are, "requires", "dependencies" and "project.optional-dependencies". Am attaching the section here. Are these the right places where the changes should be made? If not, let me know what else I should be changing. 

[build-system]
requires = [
    "hatchling>=0.13.0",
    "setuptools<61",
    "fiddler-client==1.0.6",
]
build-backend = "hatchling.build"

[project]
name = "datadog-fiddler"
description = "The fiddler check"
readme = "README.md"
license = {text = "BSD-3-Clause"}
keywords = [
    "datadog",
    "datadog agent",
    "datadog check",
    "fiddler",
]
authors = [
    { name = "Rajeev", email = "rajeev@fiddler.ai" },
]
classifiers = [
    "Development Status :: 5 - Production/Stable",
    "Intended Audience :: Developers",
    "Intended Audience :: System Administrators",
    "License :: OSI Approved :: BSD License",
    "Programming Language :: Python :: 2.7",
    "Programming Language :: Python :: 3.8",
    "Topic :: System :: Monitoring",
]
dependencies = [
    "datadog-checks-base>=25.1.0",
    "fiddler-client==1.0.6",
]
dynamic = [
    "version",
]

[project.optional-dependencies]
deps = [
    "fiddler-client==1.0.6",
]

Reply:
Add a new file `fiddler/requeriments.in` with content:
`fiddler-client==1.0.6`

In file `fiddler/pyproject.toml`:
```
[project.optional-dependencies]
deps = [
    "fiddler-client==1.0.6",
]
```
just needed

Reply:
@jose-manuel-almaza - I added requirements.in with the fiddler-client version in it. I also have the change in the pyproject.toml file. It still gives me the same error. Can you try building this on your end and see what else I might be missing? Thanks.

Reply:
Hi @rajeevthefiddler
I have done that and getting this error:

<img width="1512" alt="error" src="https://user-images.githubusercontent.com/106235809/178675012-7d384e93-782a-405b-a853-b8edc64fd4c0.png">

Reply:
@jose-manuel-almaza thank you for trying it out. And yes the error is different and I can see that you have been able to get past the error that I am seeing in my environment. The AUTH_TOKEN missing is a valid error because I have not provided the AUTH_TOKEN in that file. 

Now the question is, how did you get past the error that I am having. I was trying to debug that yesterday and today and have not made much progress. Do you think you can jump on a quick 30 minute call with me sometime tomorrow so that we can track it down? I can share my screen and you might be able to point out what I am doing wrong here. Let me know if it works and I would be really greatful. I work out of the central timezone. I am free between 9-11am and 3-5pm central time tomorrow. Let me know if you have sometime tomorrow and I can send you an invite. Thanks.

Reply:
Hi @rajeevthefiddler 
Follow these steps:

1. Clone your current code in a new and clean directory with:
`git clone https://github.com/rajeevthefiddler/integrations-extras.git`
2. Move into integrations-extras directory:
`cd integrations-extras`
3. Run ```ddev config set extras `pwd` ```
4. Run `ddev -e test -k test_check fiddler`
5. Now you should get your current error: `E   ModuleNotFoundError: No module named 'fiddler'`
6. Update the [build-system] and [project.optional-dependencies] blocks in `pyproject.toml` to be as follows:
```
[build-system]
requires = [
    "hatchling>=0.13.0",
    "setuptools<61",
]
```
```
[project.optional-dependencies]
deps = [
    "fiddler-client==1.0.6"
]
```
7. Run `rm -rf fiddler/.tox` to delete tox caches and intermediate files.
8. Run `ddev -e test -k test_check fiddler` again
9. Now you should get this new error: `E   NameError: name 'Timeout' is not defined`

Reply:
@jose-manuel-almaza , @anawishnoff - any update on this? How do I get over the "ddev test fiddler" issue that I am seeing above? 

I also had another question. Once the above steps complete, I guess I can create a wheel and then install the wheel with the DD agent. Once the wheel is installed properly, will I see my fiddler check in the /opt/datadog-agent/bin/agent/dist/checks folder?  And once the DD agent is up and running, will I see the logging that I have in my check show up in the agent.log file?

If you can give me some directions here that would be great. Thanks.

Reply:
@yzhan289 thanks for that information. I tried multiple things here to get over the first problem by adding in aggregator.assert_metric('accuracy') and aggregator.assert_metric("metric") and aggregator.assert_metric(). Tried all the above combinations in the test_check() method.

Reply:
From the exception you are getting,

```
E AssertionError: Needed at least 1 candidates for 'accuracy', got 0
E Expected:
E MetricStub(name='accuracy', type=None, value=None, tags=None, hostname=None, device=None, flush_first_value=None)
E Similar submitted:
E Score Most similar
E 0.67 MetricStub(name='fiddler.accuracy', type=0, value=0.6, tags=['model:fraud_detection_model_v1', 'project:fraud_detection'], hostname='', device=None, flush_first_value=False)
E 0.67 MetricStub(name='fiddler.accuracy', type=0, value=0.4, tags=['model:fraud_detection_model_v1', 'project:fraud_detection'], hostname='', device=None, flush_first_value=False)
E 0.67 MetricStub(name='fiddler.accuracy', type=0, value=0.0, tags=['model:fraud_detection_model_v1', 'project:fraud_detection'], hostname='', device=None, flush_first_value=False)
E 0.67 MetricStub(name='fiddler.accuracy', type=0, value=0.0, tags=['model:fraud_detection_model_v1', 'project:fraud_detection'], hostname='', device=None, flush_first_value=False)
E assert False
```
You need to be asserting for `fiddler.accuracy` rather than just `accuracy`. The correct assertion is `aggregator.assert_metric('fiddler.accuracy')`

Reply:
@yzhan289 - ok that was good. That worked.

Reply:
@yzhan289 - Andrew, For your feedback on the second error: 

"This error is because you are asserting that the service check should return you CRITICAL, but in actuality it is returning OK. You will need to modify your code to make sure that the right service check is being emitted."

My code can return an OK or a CRITICAL depending on how the code performed. So the way I have written it is if there is an exception, I will return a CRITICAL. If it passes all, then it returns an OK. Is that not what I should be doing?

Reply:
That logic sounds correct to me. However, you should look through your logic to debug why you are still getting an OK service check when you are expecting a CRITICAL. 

Looking at `test_emits_critical_service_check_when_service_is_down()`, there is no logic difference compared to the successful `test_check()`. You should send an instance of the check that actually does not exist instead. We have [many](https://github.com/DataDog/integrations-core/blob/master/druid/tests/test_unit.py) [examples](https://github.com/DataDog/integrations-extras/blob/master/cert_manager/tests/test_cert_manager.py#L75) from other integrations that simulate a failing check. Please take a look at them to debug your issue.

Reply:
@yzhan289 thanks Andrew. Ok, I am supposed to write the test code to check all situations in that method. I did not get that. I will do that. 

I had another question - once I get through these tests and create a wheel, and once I deploy the wheel in my environment with a DD agent; is there something that I can check quickly to make sure that the integration has been deployed properly? And when the agent is executing, all of the print statements that I have in my integration, will it show up in the agent.log file?

Reply:
If your Fiddler software has a Docker image, you can use [`dd_environment()`](https://github.com/DataDog/integrations-core/blob/master/arangodb/tests/conftest.py#L18-L22) to start up an instance of your integration so that running `ddev env start fiddler py38` will [start](https://datadoghq.dev/integrations-core/ddev/cli/#ddev-env-start) a new Docker container with your Fiddler software and another Docker container with an Agent to collect metrics from the Fiddler container. Your print statements should be changed to `self.log.debug()` statements if you want them to show up in the log file. After starting the Fiddler ddev environment, you can run [`ddev env check fiddler -l debug`](https://datadoghq.dev/integrations-core/ddev/cli/#ddev-env-check) to generate debug logs when running a check run.

Reply:
In my case, I wouldnt need the fiddler container for test, because the DD agent is going to talk directly to a fiddler saas endpoint. So I created the wheel and changed the print statements into self.log.debug() and when I did the ddev env check command, I got the following response. 

(base) rajeev@Rajeevs-MBP logs % ddev env check fiddler -l debug
No active environments found for `fiddler`.
See what is available to start via `ddev env ls fiddler`.
(base) rajeev@Rajeevs-MBP logs % ddev env ls fiddler
fiddler:
    py38

Let me know what my next step should be. I do see the fiddler files inside my /opt/datadog-agent folder as you can see below:

(base) rajeev@Rajeevs-MBP fiddler % pwd
/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/fiddler
(base) rajeev@Rajeevs-MBP fiddler % ls -ltr
total 32
-rw-r--r--  1 root  wheel    22 Jul 22 15:07 __about__.py
-rw-r--r--  1 root  wheel   110 Jul 22 15:07 __init__.py
-rw-r--r--  1 root  wheel  6842 Jul 22 15:07 check.py
drwxr-xr-x  3 root  wheel    96 Jul 22 15:07 data
drwxr-xr-x  5 root  wheel   160 Jul 22 15:07 __pycache__
drwxr-xr-x  8 root  wheel   256 Jul 22 15:07 config_models
(base) rajeev@Rajeevs-MBP fiddler %

Reply:
@yzhan289 - Andrew, can you give me some pointers in why my fiddler code is not getting executed after I install the wheel in my environment?

Reply:
@yzhan289 - Andrew, I finally got the ddev test fiddler to pass the two tests. The only error that shows up seems to be a style flake8 error, which does not tell me what the problem is. Here is the final output of the ddev test fiddler. Can I ignore this error and try creating the wheel and test the integration? Let me know. 

(base) rajeev@Rajeevs-MBP integrations-extras % ddev test fiddler
Running tests for `fiddler`
---------------------------
.package create: /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler/.tox/.package
.package installdeps: hatchling>=0.13.0, setuptools<61, fiddler-client==1.0.6
py38 create: /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler/.tox/py38
py38 installdeps: datadog-checks-base[deps]>=6.6.0, -rrequirements-dev.txt
py38 develop-inst: /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler
py38 installed: attrs==21.4.0,binary==1.0.0,boto3==1.24.39,botocore==1.27.39,cachetools==5.0.0,certifi==2022.6.15,cffi==1.15.1,charset-normalizer==2.1.0,coverage==6.4.2,cryptography==3.4.8,datadog-checks-base==26.0.0,datadog-checks-dev==16.5.2,-e git+https://github.com/rajeevthefiddler/integrations-extras@bd39967db672ffc48034d7e9d51e41a6e20f2e6a#egg=datadog_fiddler&subdirectory=fiddler,ddtrace==0.53.2,deepdiff==5.8.1,deprecation==2.1.0,fastavro==1.5.3,fiddler-client==1.0.6,flaky==3.7.0,idna==3.3,immutables==0.17,importlib-resources==5.9.0,iniconfig==1.1.1,jellyfish==0.9.0,jmespath==1.0.1,MarkupSafe==2.1.1,mock==4.0.3,numpy==1.23.1,ordered-set==4.1.0,packaging==21.3,pandas==1.4.3,pluggy==1.0.0,prometheus-client==0.14.1,protobuf==3.20.1,psutil==5.9.1,py==1.11.0,py-cpuinfo==8.0.0,pyarrow==8.0.0,pycparser==2.21,pydantic==1.9.0,pygal==3.0.0,pygaljs==1.0.2,pyparsing==3.0.9,pytest==7.1.2,pytest-benchmark==3.4.1,pytest-cov==3.0.0,pytest-mock==3.8.2,python-dateutil==2.8.2,pytz==2022.1,PyYAML==5.4.1,requests==2.28.1,requests-toolbelt==0.9.1,requests-unixsocket==0.3.0,s3transfer==0.6.0,simplejson==3.17.6,six==1.16.0,tenacity==8.0.1,tomli==2.0.1,typing_extensions==4.3.0,uptime==3.0.1,urllib3==1.26.11,Werkzeug==2.2.1,wrapt==1.14.0,zipp==3.8.1
py38 run-test-pre: PYTHONHASHSEED='4075251678'
py38 run-test: commands[0] | pytest -v
======================================================================================== test session starts ========================================================================================
platform darwin -- Python 3.8.8, pytest-7.1.2, pluggy-1.0.0 -- /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler/.tox/py38/bin/python
cachedir: .tox/py38/.pytest_cache
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler
plugins: benchmark-3.4.1, datadog-checks-dev-16.5.2, ddtrace-0.53.2, flaky-3.7.0, mock-3.8.2, cov-3.0.0
collected 2 items                                                                                                                                                                                   

tests/test_fiddler.py::test_check PASSED                                                                                                                                                      [ 50%]
tests/test_fiddler.py::test_emits_critical_service_check_when_service_is_down PASSED                                                                                                          [100%]

======================================================================================== 2 passed in 24.79s =========================================================================================
style create: /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler/.tox/style
style installdeps: flake8==4.0.1, flake8-bugbear==22.1.11, flake8-logging-format==0.6.0, black==22.3.0, isort==5.10.1, pydantic==1.8.2, types-PyYAML==5.4.10, types-python-dateutil==2.8.2, types_requests==2.25.11, types_six==1.16.2, types-simplejson==3.17.5
style installed: attrs==21.4.0,black==22.3.0,click==8.1.3,flake8==4.0.1,flake8-bugbear==22.1.11,flake8-logging-format==0.6.0,isort==5.10.1,mccabe==0.6.1,mypy-extensions==0.4.3,pathspec==0.9.0,platformdirs==2.5.2,pycodestyle==2.8.0,pydantic==1.8.2,pyflakes==2.4.0,tomli==2.0.1,types-python-dateutil==2.8.2,types-PyYAML==5.4.10,types-requests==2.25.11,types-simplejson==3.17.5,types-six==1.16.2,typing_extensions==4.3.0
style run-test-pre: PYTHONHASHSEED='4075251678'
style run-test: commands[0] | flake8 --config=../.flake8 .
./datadog_checks/fiddler/check.py:27:1: E302 expected 2 blank lines, found 1
./datadog_checks/fiddler/check.py:88:21: B001 Do not use bare `except:`, it also catches unexpected events like memory errors, interrupts, system exit, and so on.  Prefer `except Exception:`.  If you're sure what you're doing, be explicit and write `except BaseException:`.
ERROR: InvocationError for command /Users/rajeev/junk/integrations-extras-Jul-28-2022/integrations-extras/fiddler/.tox/style/bin/flake8 --config=../.flake8 . (exited with code 1)
______________________________________________________________________________________________ summary ______________________________________________________________________________________________
  py38: commands succeeded
ERROR:   style: commands failed

Failed!

Reply:
You can fix this style error running `ddev test fiddler -fs`

Reply:
@FlorentClarret - I did run the ddev test fiddler -fs and then tried the ddev test fiddler and you can see the outputs here. It does not look like it fixed it. The error says I might need to manually fix it, but it does not tell me where the problem is. Can I ignore the style error and proceed? 

(base) rajeev@Rajeevs-MacBook-Pro integrations-extras % ddev test fiddler -fs
Running the code formatter for `fiddler`
----------------------------------------
format_style installed: black==22.3.0,click==8.1.3,flake8==4.0.1,isort==5.10.1,mccabe==0.6.1,mypy-extensions==0.4.3,pathspec==0.9.0,platformdirs==2.5.2,pycodestyle==2.8.0,pyflakes==2.4.0,tomli==2.0.1,typing_extensions==4.3.0
format_style run-test-pre: PYTHONHASHSEED='3335614685'
format_style run-test: commands[0] | isort . --settings-path ../pyproject.toml
Skipped 1 files
format_style run-test: commands[1] | black . --config ../pyproject.toml
reformatted datadog_checks/fiddler/check.py

All done! ✨ 🍰 ✨
1 file reformatted, 12 files left unchanged.
format_style run-test: commands[2] | python -c 'print('"'"'\n[NOTE] flake8 may still report style errors for things black cannot fix, these will need to be fixed manually.'"'"')'

[NOTE] flake8 may still report style errors for things black cannot fix, these will need to be fixed manually.
format_style run-test: commands[3] | flake8 --config=../.flake8 .
______________________________________________________________________________________________ summary ______________________________________________________________________________________________
  format_style: commands succeeded
  congratulations :)

Passed!
(base) rajeev@Rajeevs-MacBook-Pro integrations-extras % ddev test fiddler    
Running tests for `fiddler`
---------------------------
/Users/rajeev/junk/integrations-extras-Jul-29-2022/integrations-extras/fiddler/.tox/py38/lib/python3.8/site-packages/setuptools/config/_apply_pyprojecttoml.py:196: UserWarning: `install_requires` overwritten in `pyproject.toml` (dependencies)
  warnings.warn(msg)
py38 develop-inst-nodeps: /Users/rajeev/junk/integrations-extras-Jul-29-2022/integrations-extras/fiddler
py38 installed: attrs==22.1.0,binary==1.0.0,boto3==1.24.40,botocore==1.27.40,cachetools==5.0.0,certifi==2022.6.15,cffi==1.15.1,charset-normalizer==2.1.0,coverage==6.4.2,cryptography==3.4.8,datadog-checks-base==26.0.0,datadog-checks-dev==16.5.2,-e git+https://github.com/rajeevthefiddler/integrations-extras@bd39967db672ffc48034d7e9d51e41a6e20f2e6a#egg=datadog_fiddler&subdirectory=fiddler,ddtrace==0.53.2,deepdiff==5.8.1,deprecation==2.1.0,fastavro==1.5.3,fiddler-client==1.0.6,flaky==3.7.0,idna==3.3,immutables==0.17,importlib-resources==5.9.0,iniconfig==1.1.1,jellyfish==0.9.0,jmespath==1.0.1,MarkupSafe==2.1.1,mock==4.0.3,numpy==1.23.1,ordered-set==4.1.0,packaging==21.3,pandas==1.4.3,pluggy==1.0.0,prometheus-client==0.14.1,protobuf==3.20.1,psutil==5.9.1,py==1.11.0,py-cpuinfo==8.0.0,pyarrow==8.0.0,pycparser==2.21,pydantic==1.9.0,pygal==3.0.0,pygaljs==1.0.2,pyparsing==3.0.9,pytest==7.1.2,pytest-benchmark==3.4.1,pytest-cov==3.0.0,pytest-mock==3.8.2,python-dateutil==2.8.2,pytz==2022.1,PyYAML==5.4.1,requests==2.28.1,requests-toolbelt==0.9.1,requests-unixsocket==0.3.0,s3transfer==0.6.0,simplejson==3.17.6,six==1.16.0,tenacity==8.0.1,tomli==2.0.1,typing_extensions==4.3.0,uptime==3.0.1,urllib3==1.26.11,Werkzeug==2.2.1,wrapt==1.14.0,zipp==3.8.1
py38 run-test-pre: PYTHONHASHSEED='3483669214'
py38 run-test: commands[0] | pytest -v
======================================================================================== test session starts ========================================================================================
platform darwin -- Python 3.8.8, pytest-7.1.2, pluggy-1.0.0 -- /Users/rajeev/junk/integrations-extras-Jul-29-2022/integrations-extras/fiddler/.tox/py38/bin/python
cachedir: .tox/py38/.pytest_cache
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/rajeev/junk/integrations-extras-Jul-29-2022/integrations-extras/fiddler
plugins: benchmark-3.4.1, datadog-checks-dev-16.5.2, ddtrace-0.53.2, flaky-3.7.0, mock-3.8.2, cov-3.0.0
collected 2 items                                                                                                                                                                                   

tests/test_fiddler.py::test_check PASSED                                                                                                                                                      [ 50%]
tests/test_fiddler.py::test_emits_critical_service_check_when_service_is_down PASSED                                                                                                          [100%]

========================================================================================= 2 passed in 5.77s =========================================================================================
style installed: attrs==22.1.0,black==22.3.0,click==8.1.3,flake8==4.0.1,flake8-bugbear==22.1.11,flake8-logging-format==0.6.0,isort==5.10.1,mccabe==0.6.1,mypy-extensions==0.4.3,pathspec==0.9.0,platformdirs==2.5.2,pycodestyle==2.8.0,pydantic==1.8.2,pyflakes==2.4.0,tomli==2.0.1,types-python-dateutil==2.8.2,types-PyYAML==5.4.10,types-requests==2.25.11,types-simplejson==3.17.5,types-six==1.16.2,typing_extensions==4.3.0
style run-test-pre: PYTHONHASHSEED='3483669214'
style run-test: commands[0] | flake8 --config=../.flake8 .
./datadog_checks/fiddler/check.py:91:21: B001 Do not use bare `except:`, it also catches unexpected events like memory errors, interrupts, system exit, and so on.  Prefer `except Exception:`.  If you're sure what you're doing, be explicit and write `except BaseException:`.
ERROR: InvocationError for command /Users/rajeev/junk/integrations-extras-Jul-29-2022/integrations-extras/fiddler/.tox/style/bin/flake8 --config=../.flake8 . (exited with code 1)
______________________________________________________________________________________________ summary ______________________________________________________________________________________________
  py38: commands succeeded
ERROR:   style: commands failed

Failed!